### PR TITLE
Allow to publish local address to local zookeeper only

### PR DESCRIPTION
**Pull request by jitendradhawan**

### CODE REVIEW

File: src/main/java/io/appform/dropwizard/discovery/bundle/ServiceDiscoveryBundle.java [hunk 1]
```diff
@@ -185,11 +187,16 @@ protected int getPort(T configuration) {
     }
 
     protected String getHost() throws UnknownHostException {
-        val host = serviceDiscoveryConfiguration.getPublishedHost();
+        String host = serviceDiscoveryConfiguration.getPublishedHost();
```
Comment (santanusinha) [inline src/main/java/io/appform/dropwizard/discovery/bundle/ServiceDiscoveryBundle.java:190]: Move to val and use ternary operator here

File: src/main/java/io/appform/dropwizard/discovery/bundle/Constants.java [hunk 2]
```diff
@@ -29,4 +31,6 @@ public class Constants {
     public static final int DEFAULT_PORT = -1;
     public static final int DEFAULT_DW_CHECK_INTERVAL = 15;
     public static final int DEFAULT_RETRY_CONN_INTERVAL = 5000;
+
+    public static final List<String> LOCALHOSTS  = Arrays.asList("127.0.0.1", "localhost");
```
Comment (santanusinha) [inline src/main/java/io/appform/dropwizard/discovery/bundle/Constants.java:35]: LOCAL_ADDRESSES

File: src/main/java/io/appform/dropwizard/discovery/bundle/ServiceDiscoveryBundle.java [hunk 3]
```diff
@@ -185,11 +187,16 @@ protected int getPort(T configuration) {
...
         if (Strings.isNullOrEmpty(host) || host.equals(Constants.DEFAULT_HOST)) {
-            return InetAddress.getLocalHost()
+            host = InetAddress.getLocalHost()
                     .getCanonicalHostName();
         }
+        Preconditions.checkArgument(
+                !LOCALHOSTS.contains(host) ||
+                        LOCALHOSTS.stream().anyMatch(localhost -> serviceDiscoveryConfiguration.getZookeeper().contains(localhost)),
+                "Looks like publishedHost has been set to localhost/127.0.0.1 and zookeeper has not been set to localhost/127.0.0.1. This is wrong. \n" +
+                        "Set zookeeper host to localhost/127.0.0.1 in config in order to set publishedHost as localhost/127.0.0.1 for the running service");
```
Comment (santanusinha) [inline src/main/java/io/appform/dropwizard/discovery/bundle/ServiceDiscoveryBundle.java:199]: The check here will need to be much more involved for situations where someone has set some hostname for localhost for whatever reason.

**Review by santanusinha — CHANGES_REQUESTED** [review]

File: src/main/java/io/appform/dropwizard/discovery/bundle/ServiceDiscoveryBundle.java (hunk 3 above)
Comment (jitendradhawan) [reply src/main/java/io/appform/dropwizard/discovery/bundle/ServiceDiscoveryBundle.java]: only allowing publishedHost as local address with zookeeper pointed to  local address. Please suggest more precondition checks

File: src/main/java/io/appform/dropwizard/discovery/bundle/Constants.java [hunk 4]
```diff
@@ -29,4 +31,9 @@ public class Constants {
     public static final int DEFAULT_PORT = -1;
     public static final int DEFAULT_DW_CHECK_INTERVAL = 15;
     public static final int DEFAULT_RETRY_CONN_INTERVAL = 5000;
+
+    public static final String ZOOKEEPER_HOST_DELIMITER = ",";
+    public static final String HOST_PORT_DELIMITER = ":";
+
+    public static final List<String> LOCAL_ADDRESSES = Arrays.asList("127.0.0.1","127.0.1.1", "localhost");
```
Comment (santanusinha) [inline src/main/java/io/appform/dropwizard/discovery/bundle/Constants.java:38]: This should be a Set

File: src/main/java/io/appform/dropwizard/discovery/bundle/ServiceDiscoveryBundle.java (hunk 3 above)
Comment (jitendradhawan) [reply src/main/java/io/appform/dropwizard/discovery/bundle/ServiceDiscoveryBundle.java]: Have put check for custom hostnames configured for localhost, 127.0.0.1, 127.0.1.1

File: pom.xml [hunk 5]
```diff
@@ -272,6 +273,11 @@
             <version>${logback.version}</version>
             <scope>test</scope>
         </dependency>
+        <dependency>
+            <groupId>com.alibaba</groupId>
+            <artifactId>dns-cache-manipulator</artifactId>
+            <version>${dns.cache.manipulator.version}</version>
```
[the remark this answers is not in the document]
Comment (jitendradhawan) [reply pom.xml:279]: done

File: src/main/java/io/appform/dropwizard/discovery/bundle/ServiceDiscoveryBundle.java [hunk 6]
```diff
@@ -188,11 +193,36 @@ protected int getPort(T configuration) {
...
+                                .getHostAddress();
+                    } catch (UnknownHostException e) {
+                        throw new RuntimeException(
+                                String.format("Couldn't resolve host address for zkHost : %s", zkHost), e);
+                    }
+                })
+                .collect(Collectors.toSet());
+
+        Preconditions.checkArgument(
+                !LOCAL_ADDRESSES.contains(publishedHostAddress) || new HashSet<>(LOCAL_ADDRESSES).containsAll(
```
[the remark this answers is not in the document]
Comment (jitendradhawan) [reply src/main/java/io/appform/dropwizard/discovery/bundle/ServiceDiscoveryBundle.java]: done

**Pull request closed without merging**